### PR TITLE
Recovery Service Vault: Enable Immutability Settings

### PR DESCRIPTION
**Pull request by v-rahkuma**
Recovery Service Vault: Enable Immutability Settings

### CODE REVIEW

**Comment by vaishnavipulluri** [comment]
"Subscription" casing should be taken care

**Comment by vaishnavipulluri** [comment]
Recheck on the permissions required to run the script and accordingly tweak the messages

File: Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1 [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+    $controlIds = "Azure_RecoveryServicesVault_DP_Enable_Immutability"
+
+    if ($AutoRemediation) {
+        if (-not (Test-Path -Path $Path)) {    
+            Write-Host "File containing failing controls details [$($Path)] not found. Skipping remediation..." -ForegroundColor $([Constants]::MessageType.Error)    
+            Write-Host $([Constants]::DoubleDashLine)    
+            return    
+        }
+        Write-Host "Fetching all Recovery Services Vault(s) failing for the [$($controlIds)] control from [$($Path)]..." -ForegroundColor $([Constants]::MessageType.Info)
+        Write-Host $([Constants]::SingleDashLine)
```
Comment (vaishnavipulluri) [inline Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1:242]: Single dash line post fetching messages why?

File: Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1 [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+                $logSkippedResources += $logResource
+
+                Write-Host $([Constants]::SingleDashLine)
+                Write-Host "Error fetching Recovery Services Vault resource: Resource ID: [$($_.ResourceId)]. Error: $($_)" -ForegroundColor $([Constants]::MessageType.Error)
+            }    
+        }    
+    }
+    else {
+        if ([String]::IsNullOrWhiteSpace($FilePath)) {
+            Write-Host "Fetching all Recovery Services Vault(s) in Subscription: [$($context.Subscription.SubscriptionId)]" -ForegroundColor $([Constants]::MessageType.Info)
```
Comment (vaishnavipulluri) [inline Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1:277]: Should end with 3 dots

File: Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1 [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+        return
+    }
+
+    Write-Host "Found [$($totalRecoveryServiceVaults)] Recovery Services Vault(s)." -ForegroundColor $([Constants]::MessageType.Update)
+    Write-Host $([Constants]::SingleDashLine)
+    
+    # Recovery Services Vault(s) where immutability is not locked
+    $RecoveryServiceVaultsWithoutImmutability = @()
+
+    Write-Host "Separating Recovery Services Vault(s) for which immutability is not locked..." -ForegroundColor $([Constants]::MessageType.Info)
```
Comment (vaishnavipulluri) [inline Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1:336]: We are communicating just about "not locked" state, but control would look for both enabled and locked right. Please tweak the messaging accordingly.

**Comment by vaishnavipulluri** [comment]
Note added speaks about rollback please recheck on it.

File: Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1 (hunk 1 above)
Comment (v-rahkuma) [reply Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1]: Removed Single dash line

File: Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1 (hunk 3 above)
Comment (v-rahkuma) [reply Scripts/RemediationScripts/Remediate-EnableImmutabilitySettingForRecoveryServiceVault.ps1:336]: Done

**Comment by v-rahkuma** [comment]
> Recheck on the permissions required to run the script and accordingly tweak the messages

Done,

**Comment by v-rahkuma** [comment]
> Note added speaks about rollback please recheck on it.

As discussed, marking this comment as closed, as that Note is a part of auto remediation script, which remediates  multiple controls at once

**Review by Aboli-msft — APPROVED** [review]

**Review by vaishnavipulluri — APPROVED** [review]

**Pull request merged**